go/src/url_shortener/stats.go

```diff
@@ -0,0 +1,98 @@
+package main
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+)
+
+// StatsJSON Statistic data ready for JSON serialization
+type StatsJSON struct {
+	ServerStats serverStatsJSON `json:"server_stats"`
+
+	mux sync.Mutex
+}
+
+type serverStatsJSON struct {
+	TotalURL  int           `json:"total_url"`
+	Redirects redirectsJSON `json:"redirects"`
+	Handlers  []handlerJSON `json:"handlers"`
+}
+
+type redirectsJSON struct {
+	Success int `json:"success"`
+	Failed  int `json:"failed"`
+}
+
+type handlerJSON struct {
+	Name  string `json:"name"`
+	Count int    `json:"count"`
+}
+
+func (s *StatsJSON) String() string {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	stats := s.ServerStats
+
+	pairsInCache := fmt.Sprintf("Number of long/short URL pairs: %v", stats.TotalURL)
+
+	redirects := stats.Redirects
+	succeededRedirects := fmt.Sprintf("Succeeded redirects: %v", redirects.Success)
+	failedRedirects := fmt.Sprintf("Failed redirects: %v", redirects.Failed)
+
+	handlers := &stats.Handlers
+	handlerCalls := make([]string, 0, len(*handlers))
+
+	for _, handler := range *handlers {
+		handlerCalls = append(handlerCalls, fmt.Sprintf("Handler %s called %v time(s)", handler.Name, handler.Count))
+	}
+
+	statsBody := fmt.Sprintf("Some statistics:\n\n%s\n%s\n%s\n%s", pairsInCache, succeededRedirects, failedRedirects, strings.Join(handlerCalls, "\n"))
+	return statsBody
+}
+
+func (s *StatsJSON) updateTotalURL(totalURL int) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	stats := &s.ServerStats
+
+	stats.TotalURL = totalURL
+}
+
+func (s *StatsJSON) incrementHandlerCounter(handlerURL string, succeeded bool) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	stats := &s.ServerStats
+
+	handlers := &stats.Handlers
+	if handlers == nil {
+		*handlers = make([]handlerJSON, 0)
+	}
+
+	found := false
+	for i := range *handlers {
+		handler := &(*handlers)[i]
+		name := handler.Name
+		if name != handlerURL {
+			continue
+		}
+
+		handler.Count++
+		found = true
+		break
+	}
+
+	if !found {
+		*handlers = append(*handlers, handlerJSON{handlerURL, 1})
+	}
+
+	redirects := &stats.Redirects
+	if succeeded {
+		redirects.Success++
+	} else {
+		redirects.Failed++
+	}
+}
```

Comment:
I think this function could be rewritten in a much simpler way:
 - constructors are not part of the language, but as soon as a `struct` has a non-trivial initialization the convention is to write what we call a _constructor function_. It simply is a function of the form: `newMyStruct(...) *MyStruct`  or  `newMyStruct(...) MyStruct`  depending on the struct semantics. If such a function existed for `StatsJSON`, `stats.Handlers` could be allocated there. Also, since in our case the total number of handlers is known beforehand, all the handlers could be allocated in the constructor.
 - again, since the handlers are known beforehand, we could assign an index to each handler (using the `const` and `iota` keywords) and pass that index to `incrementHandlerCounter` rather than the string. Doing so would also make it more easy to update the number of succeeded/failed redirects atomically

Reply:
I made this, but it seems to me that stats and server are coupled again.